### ISSUE TITLE
Add optional way of specifying to compile with STAN_OPENCL or STAN_THREADS

### ISSUE THREAD

**Issue by rok-cesnovar** [stan-dev/cmdstanr#29]
In order to support OpenCL execution we need to write some flags to the make/local before compiling. Same goes for STAN_THREADS when using map_rect. The same idea could then be used for STAN_MPI but I dont have much experience with that to test it so lets gor for the first 2 for now.

I am going to try to implement it this week after the JSON PR. For now I would just like some feedback if this is reasonable:

- add arguments to the compile function. It would be something like `model$compile(stan_opencl = TRUE, opencl_platform_id = 0, opencl_device_id = 1)` for OpenCL and `model$compile(stan_threads = TRUE)`. The names of the argument are up for debate. I used the same names as they are used in the make/local file. Before compiling we output the flags to make/local, compile and then delete the make/local so we dont effect any subsequent model compilation.

- add a way of speficying the number of threads used when using `stan_threads` (STAN_NUM_THREADS is an environment variable so that can not be handled at compile time). For that I propose to add `model$set_num_threads()` and `model$num_threads()`. The former sets the environment variable and the latter one reads the environment variable.

I am pushing for some of this performance stuff (JSON & OpenCL) because we plan on doing a "public request for performance testing" some time after the 2.21 release for the newly realesed GPU GLM support and I would really like to use cmdstanr instead of our custom R scripts.

**Comment by jgabry** [comment]
That all sounds great! For the names, I think they're good, but I would slightly prefer changing these

```r
stan_opencl = TRUE
stan_threads = TRUE
```
to these
```r
opencl = TRUE
threading = TRUE
```
The `stan_` prefix seems redundant, right?

**Comment by jgabry** [comment]
Closing now that #34 is merged

**Comment by stephensrmmartin** [comment]
Not sure this is worth a bug report, but if I use:

```
sm_nocl <- cmdstan_model("../Stan/rbf.stan", force = TRUE, cpp_options = list(stan_opencl = FALSE, stan_threads = TRUE))

```
then I still get an opencl-enabled model. I have to drop `stan_opencl` from the cpp_options in order to have a threads-only enabled model. Is that intended behavior?

**Comment by rok-cesnovar** [comment]
> but it was intended.

That should check that `stan_opencl` is also TRUE. Thanks for finding it.

**Comment by jgabry** [comment]
Should be fixed by #491. Thanks @stephensrmmartin for pointing this out.
